### DISCUSSION TITLE
Terraform change bios

### DISCUSSION THREAD

**Discussion by kalisteel** [bpg/terraform-provider-proxmox#1214] [Q&A]
Terraform change bios on "seabios" but I need "ovmf"
Plan terraform:
```
-/+ resource "proxmox_virtual_environment_vm" "vm" {
      ~ bios                    = "ovmf" -> "seabios"
      ~ id                      = "108" -> (known after apply)
      ~ ipv4_addresses          = [] -> (known after apply)
      ~ ipv6_addresses          = [] -> (known after apply)
      + keyboard_layout         = "en-us"
      ~ mac_addresses           = [] -> (known after apply)
      + migrate                 = false
        name                    = "vm-vk-loki"
      ~ network_interface_names = [] -> (known after apply)
      + on_boot                 = true
      + reboot                  = false
      + stop_on_destroy         = false
        tags                    = []
      + timeout_clone           = 1800
      + timeout_create          = 1800
      + timeout_migrate         = 1800
      + timeout_move_disk       = 1800
      + timeout_reboot          = 1800
      + timeout_shutdown_vm     = 1800
      + timeout_start_vm        = 1800
      + timeout_stop_vm         = 300
        # (7 unchanged attributes hidden)
      ~ agent {
          + type    = "virtio"
            # (3 unchanged attributes hidden)
        }
      ~ cpu {
          - flags        = [] -> null
          ~ type         = "qemu64" -> "host"
            # (7 unchanged attributes hidden)
        }
      ~ disk {
          ~ cache             = "none" -> "writeback"
          ~ discard           = "ignore" -> "on"
          ~ path_in_datastore = "vm-108-disk-0" -> (known after apply)
          ~ size              = 0 -> 20
            # (5 unchanged attributes hidden)
        }
      - efi_disk { # forces replacement
          - datastore_id      = "local-lvm" -> null
          - file_format       = "raw" -> null
          - pre_enrolled_keys = true -> null # forces replacement
          - type              = "4m" -> null # forces replacement
        }
      ~ initialization {
          - interface    = "ide0" -> null
            # (1 unchanged attribute hidden)
            # (1 unchanged block hidden)
        }
      ~ network_device {
          ~ mac_address = "BC:24:11:CC:88:69" -> (known after apply)
            # (8 unchanged attributes hidden)
        }
      - vga {
          - enabled = true -> null
          - memory  = 0 -> null
        }
        # (2 unchanged blocks hidden)
    }
```
main.tf
```
      name        = "vm-vk-loki"
      vm_id       = 108
      dns_enabled = true
      dns_zone    = "staging.fun-box"

      initialization_datastore_id = "local-lvm"
      initialization_ipv4_address = "10.0.2.244/24"
      initialization_ipv4_gateway = "10.0.2.1"

      scsi_hardware = "pvscsi"
      node_name     = "pve-mmts9-node1"
      
      cpu_cores        = 2
      dedicated_memory = 4096

      bios = "ovmf"

      disks = [
        {
          interface_type = "scsi"
          size           = "20"
          datastore_id   = "local-lvm"
        },
      ]

      network_device_vlan_id = "326"
    },
```
Mine provider
```
      source  = "bpg/proxmox"
      version = "0.48.1"
```
I can't change bios on virtual machine, because I don't want to lose the data on it.I'm moving the VM from Vsphere(ESXi)

I'm try change the code in module and add string: bios = lookup(each.value, "ovmf", "seabios")

but I realized that it is not necessary, because the provider has a default value.

**Comment by bpg-dev** [comment]
Hi @kalisteel 👋🏼 Could you provide a bit more detail about the flow that led to this result?
- Did you create the VM from the provided `main.tf` template initially, and then call `terraform apply` again with the same template?
- Or did you import the VM resource into Terraform state and then attempt to build a template for it?
- If you imported it, how did you do so? What commands did you use?

**Reply by kalisteel** [reply]
I transferred vm from vmware to proxmox using ovftool, then set up the proxmox machine inside and ran the terraform plan by first creating a machine in the main.tf. After that, the plan shows that the bios is being replaced. I did the terraform operations through gitlab.

**Reply by bpg-dev** [reply]
Sorry, I still don't fully understand your process.
> .., then set up the proxmox machine inside 

How did you do that? Manually via the Proxmox UI?

**Reply by bpg-dev** [reply]
Also, the `main.tf` you've posted doesn't seem to be correct (or it came from a different provider). Most of attributes referenced there [do not exist](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm).

> I did the terraform operations through gitlab.

Could you please double check what is the plan file the GitLab flow is using? Also, I'll be curious what is the current TF state cached on GitLab. It might be corrupted somehow by replacing providers (if that was the case), so the plan apply may give you inconsistent results.

**Reply by kalisteel** [reply]
pardon my English. Yes, I configure in the web interface of proxmox. I don't know exactly what provider the code is from, it was already done before me and I continue to do it according to the template, the other () work, but on oracle linux it changes the bios, I can send the full main.tf.
The code specifies which provider to refer to,I haven't changed suppliers.

**Reply by bpg-dev** [reply]
The full main.tf would help.
